chore: y axis management in metrics explorer

amlannandy opened the pull request:
## 📄 Summary

<!-- Describe the purpose of the PR in a few sentences. What does it fix/add/update? -->

- When a single query is selected and if it already has a unit, auto-populate the y axis unit selector with it

https://github.com/user-attachments/assets/21d0a56f-1a7e-4b4a-8476-d15fa2cc9cad

- When a single query is selected and If it doesn't have a unit, let the user select and save one as well

https://github.com/user-attachments/assets/098ed656-68ae-4df9-a672-72334291d885

- When multiple queries are selected and If all have the same unit, auto-populate the y axis unit selector with it.

https://github.com/user-attachments/assets/5e7bfaa9-c288-4d04-812a-0e835a3dd3cd

- When multiple queries are selected and If one of them doesn't have a set unit, showing a warning tooltip to allow them to go to metric details drawer and set it

https://github.com/user-attachments/assets/e7d32d36-8b0f-4a4e-8f05-abe74a08053b

- When multiple queries are selected and If all have different units, hide the y axis unit selector. Enable 'One Chart per Query` by default and it cannot be toggled back.

https://github.com/user-attachments/assets/b4046db4-2e5f-4f9e-ba6c-d5637db72e7a

---

## ✅ Changes

- [ ] Feature: Brief description
- [ ] Bug fix: Brief description

---

## 🏷️ Required: Add Relevant Labels

> ⚠️ **Manually add appropriate labels in the PR sidebar**  
Please select one or more labels (as applicable):

ex:

- `frontend`
- `backend`
- `devops`
- `bug`
- `enhancement`
- `ui`
- `test`

---

## 👥 Reviewers

> Tag the relevant teams for review:

- frontend / backend / devops

---

## 🧪 How to Test

<!-- Describe how reviewers can test this PR -->
1. ...
2. ...
3. ...

---

## 🔍 Related Issues

<!-- Reference any related issues (e.g. Fixes SigNoz/signoz#123, Closes SigNoz/signoz#456) -->
Closes #

---

## 📸 Screenshots / Screen Recording (if applicable / mandatory for UI related changes)

<!-- Add screenshots or GIFs to help visualize changes -->

---

## 📋 Checklist

- [ ] Dev Review
- [ ] Test cases added (Unit/ Integration / E2E)
- [ ] Manually tested the changes


---

## 👀 Notes for Reviewers

<!-- Anything reviewers should keep in mind while reviewing -->

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces unit-aware charting and metadata-backed behavior in Metrics Explorer.
> 
> - Adds `GET /v2/metrics/metadata` client, React Query key, and `useGetMultipleMetrics`/`useGetMetrics` hooks to fetch metric metadata
> - Maps metric units to universal units via `getMetricUnits` and injects per-query `unit` in `splitQueryIntoOneChartPerQuery`
> - Explorer: auto-selects y‑axis unit when all metrics share one; hides selector when multiple/different; shows selector for single metric with no unit; disables/enforces "1 chart/query" when units differ; includes `unit` in exported queries
> - TimeSeries: shows warning tooltip per chart when a metric lacks a unit with deep-link to `MetricDetails`; passes unit to `TimeSeriesView`
> - UI: adds tooltip-disabled switch logic and styles for y‑axis selector and per-chart warnings
> - MetricDetails: makes `openInspectModal` optional and shows Inspect button conditionally
> - Tests: new unit tests for Explorer, TimeSeries, and utils (unit logic, splitting, hooks)
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2f7b7e15f3a9db5bbcd2cb35cac4e1e8ecc776af. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

amlannandy commented:
@srikanthccv Now that the other PR is merged, let me replace the component here and test it again. Then I think we can review it.

amlannandy commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/Explorer.test.tsx:
```diff
@@ -188,20 +208,116 @@ describe('Explorer', () => {
+
+		const oneChartPerQueryToggle = screen.getByRole('switch');
+		expect(oneChartPerQueryToggle).toBeChecked();
+	});
```
should be checking if it's enabled or not. will fix

amlannandy commented on frontend/src/api/metricsExplorer/getMetricMetadata.ts:
```diff
@@ -0,0 +1,34 @@
+		return {
+			httpStatusCode: response.status,
+			data: response.data.data,
+		};
```
Verified that it works. Acceptable

amlannandy commented on frontend/src/hooks/metricsExplorer/useGetMultipleMetrics.ts:
```diff
@@ -0,0 +1,33 @@
+		metricNames.map(
+			(metricName) =>
+				({
+					queryKey: [REACT_QUERY_KEY.GET_METRIC_DETAILS, metricName],
```
Valid as we changed the API. Adding a new query key for it

srikanthccv commented on frontend/src/container/MetricsExplorer/Explorer/Explorer.tsx:
```diff
@@ -68,15 +135,20 @@ function Explorer(): JSX.Element {
+			};
+		}
+		return query;
+	}, [currentQuery, updateAllQueriesOperators, yAxisUnit]);
```
So as I understand, using the same example above, the export functionality exports to one panel for all queries and sets the unit as undefined?

srikanthccv commented on frontend/src/container/MetricsExplorer/Explorer/TimeSeries.tsx:
```diff
@@ -126,32 +135,160 @@ function TimeSeries({
+	// 				});
+	// 			},
+	// 		},
+	// 	);
```
Will we uncomment when working on the search bar revamp? Please add a TODO stating the reason so reader know why this is commented out.

AshwinBhatkal replied:
I would prefer that we do not have this in the code at all unless it is a fast follow up. The commit can be saved by the developer though

srikanthccv commented on frontend/src/container/MetricsExplorer/Explorer/TimeSeries.tsx:
```diff
@@ -126,32 +135,160 @@ function TimeSeries({
+
+					// Show the no unit warning if -
+					// 1. The metric query is not loading
+					// 2. The metric units are not loading
```
What happens to the time series chart if the query range API is resolved but the metadata is not resolved yet? Is that possible?

amlannandy commented:
@srikanthccv Have addressed your comments. A couple of points:

- For the newly created api func and types, I've created a new v2 folder. For splitting the existing api funcs and types to v1, v2 and common. I 've created a separate ticket. As doing them in this PR would have made the PR a bit too big. - https://github.com/SigNoz/engineering-pod/issues/3450

- For this point - [https://github.com/SigNoz/signoz/pull/9587#discussion_r2630045326](https://github.com/SigNoz/signoz/pull/9587#discussion_r2630045326) - I've done the required changes in the metrics explorer side, but the whole functionality would work properly only in the next PR where I make the changes in alerts and dashboards as well, so we can verify this change in the next PR

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/Explorer.tsx:
```diff
@@ -48,7 +80,64 @@ function Explorer(): JSX.Element {
+			setYAxisUnit(undefined);
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [JSON.stringify(units), areAllMetricUnitsSame]);
```
stringify is a code smell here

amlannandy replied:
Passing the units array in itself was causing undesired re-renders. So. I tried to stringify it. But I'll clean up the dependency array here in a better way.

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+		expect(mockSetIsMetricDetailsOpen).toHaveBeenCalledWith('metric1');
+	});
+
+	it.skip('shows Save unit button when metric had no unit but one is selected', () => {
```
why skip?

amlannandy replied:
We are skipping the tests for the save functionality as we have disabled the save feature itself. As we want to migrate the metrics explorer APIs from v1 to v2 before enabling it. 
I've a separate ticket to enable there after the migration: https://github.com/SigNoz/signoz/issues/9861

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+	.mockReturnValue({
+		mutate: mockUpdateMetricMetadata,
+		isLoading: false,
+	} as any);
```
is any required here?

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+		isLoading: false,
+	} as any);
+
+jest.mock('hooks/queryBuilder/useQueryBuilder', () => {
```
This seems like a generic override. Can it be moved to a different common file?

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+		expect(yesButton).toBeEnabled();
+	});
+
+	it.skip('clicking on save unit button shoould upated metric metadata', () => {
```
why skip

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+		});
+
+		const alertIcon = container.querySelector('.no-unit-warning') as HTMLElement;
+		fireEvent.mouseOver(alertIcon);
```
Please change at all places

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/MetricDetails/MetricDetails.tsx:
```diff
@@ -131,8 +131,8 @@ function MetricDetails({
 						</Button>
-						{/* Show the based on the feature flag. Will remove before releasing the feature */}
-						{showInspectFeature && (
+						{/* Show the inspect button if the metric type is GAUGE */}
```
This comment seems to be at the wrong place

amlannandy replied:
Actually it's at the intended place. `showInspectFeature` checks if the metric type is gauge

AshwinBhatkal commented on frontend/src/container/MetricsExplorer/Explorer/utils.tsx:
```diff
@@ -35,3 +55,43 @@ export const splitQueryIntoOneChartPerQuery = (query: Query): Query[] => {
+	isError: boolean;
+	metrics: (MetricMetadata | undefined)[];
+} {
+	const metricsData = useGetMultipleMetrics(metricNames, {
```
Isn't there a BE api just for this? We should create one if so.
We should avoid multiple calls wherever we can. This is one of those cases it feels like

amlannandy replied:
We don't have an API for fetching metadata of multiple metrics at once. Just one at a time. The API in this case should not be that heavy, so it should be fine I think.
@srikanthccv may be able to answer this better.

amlannandy commented on frontend/src/container/MetricsExplorer/Explorer/__tests__/TimeSeries.test.tsx:
```diff
@@ -0,0 +1,190 @@
+		fireEvent.mouseOver(alertIcon);
+		return waitFor(() =>
+			expect(
+				screen.getByText(/This metric does not have a unit/i),
```
Yeah I think I replaced it at places but at some it got left. Let me replace all of them

amlannandy commented on frontend/src/container/MetricsExplorer/Explorer/TimeSeries.tsx:
```diff
@@ -126,32 +137,161 @@ function TimeSeries({
+	// Show the y axis unit selector if -
+	// 1. There is only one metric
+	// 2. The metric has no saved unit
+	const showYAxisUnitSelector = useMemo(
```
Will clean up the props a bit. But the `TimeSeries.tsx` component was meant to be only used here in metrics only which is why I think it's okay for it to have specific implementations.
The generic component you are referring to it is `container/TimeSeriesView/TimeSeriesView` which we are not modifying here.
